addons/sms/controllers/discuss.py

```diff
@@ -0,0 +1,18 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import odoo.addons.mail.controllers.discuss as discuss
+from odoo.osv.expression import OR
+
+
+class DiscussController(discuss.DiscussController):
+    def _mail_thread_messages_filter(self):
+        domain = super()._mail_thread_messages_filter()
+        # Exclude sms not sent because of a permanent failure: blacklisted, duplicated or opted out
+        domain.append(
```

Comment:
I would use expression.AND instead

Reply:
Maybe I'm missing something but to use AND, I have to do it like this:
```python
return [AND([
            *super()._mail_thread_messages_filter(),
            # Exclude sms not sent because of a permanent failure: blacklisted, duplicated or opted out
            OR([
                [('sms_ids', '=', False)],  # Not all mail.message have a related sms.sms
                [('sms_ids.failure_type', 'not in', ('sms_blacklist', 'sms_optout', 'sms_duplicate'))],
            ])
        ])]
```
But the AND will anyway be applied by the class using the result of the method so it is a bit redundant (and anyway using an other operator like this would be dangerous as we don't know the order of execution of the inherited method).

I have updated the code with a shorter version but less efficient so depending on your answer I may come back to the previous one or another solution. Thanks